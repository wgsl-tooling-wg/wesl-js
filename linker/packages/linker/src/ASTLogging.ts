import { AbstractElem2, ModuleElem } from "./AbstractElems2.ts";
import { treeToString } from "./ImportTree.ts";
import { LineWrapper } from "./LineWrapper.ts";

export function astTree(elem: AbstractElem2, indent = 0): string {
  const { kind, contents } = elem as ModuleElem;
  const str = new LineWrapper(indent);
  str.add(kind);
  addElemFields(elem, str);
  let childStrings: string[] = [];
  if (contents) {
    childStrings = contents.map(e => astTree(e, indent + 2));
  }
  if (childStrings.length) {
    str.nl();
    str.addBlock(childStrings.join("\n"), false);
  }

  return str.result;
}

function addElemFields(elem: AbstractElem2, str: LineWrapper): void {
  addTextFields(elem, str) ||
    addVarishFields(elem, str) ||
    addStructFields(elem, str) ||
    addNameFields(elem, str) ||
    addFnFields(elem, str) ||
    addAliasFields(elem, str) ||
    addImport(elem, str) ||
    addIdentFields(elem, str);
}

function addVarishFields(
  elem: AbstractElem2,
  str: LineWrapper,
): true | undefined {
  const { kind } = elem;
  if (kind === "var" || kind === "const" || kind === "override") {
    str.add(" " + elem.name.ident.originalName);
    if (elem.typeRef) {
      str.add(":" + elem.typeRef.ident.originalName);
    }
    return true;
  }
}

function addTextFields(
  elem: AbstractElem2,
  str: LineWrapper,
): true | undefined {
  if (elem.kind === "text") {
    const { src, start, end } = elem;
    str.add(` '${src.slice(start, end)}'`);
    return true;
  }
}

function addIdentFields(
  elem: AbstractElem2,
  str: LineWrapper,
): true | undefined {
  if (elem.kind === "ident") {
    const { ident } = elem;
    const prefix = ident.kind === "decl" ? "%" : "";
    str.add(" " + prefix + elem.ident.originalName);
    return true;
  }
}

function addAliasFields(
  elem: AbstractElem2,
  str: LineWrapper,
): true | undefined {
  if (elem.kind === "alias") {
    const { name, typeRef } = elem;
    const prefix = name.ident.kind === "decl" ? "%" : "";
    str.add(" " + prefix + name.ident.originalName);
    str.add("=" + typeRef.ident.originalName);
    return true;
  }
}

function addStructFields(
  elem: AbstractElem2,
  str: LineWrapper,
): true | undefined {
  if (elem.kind === "struct") {
    str.add(" " + elem.name.ident.originalName);
    return true;
  }
}

function addImport(elem: AbstractElem2, str: LineWrapper): true | undefined {
  if (elem.kind === "import") {
    str.add(" " + treeToString(elem.imports));
    return true;
  }
}

function addNameFields(
  elem: AbstractElem2,
  str: LineWrapper,
): true | undefined {
  if (elem.kind === "name") {
    str.add(" " + elem.name);
    return true;
  }
}

function addFnFields(elem: AbstractElem2, str: LineWrapper): true | undefined {
  if (elem.kind === "fn") {
    const { name, params, returnType } = elem;
    str.add(" " + name.ident.originalName);

    str.add("(");
    const paramStrs = params
      .map(p => p.name.ident.originalName + ": " + p.typeRef.ident.originalName)
      .join(", ");
    str.add(paramStrs);
    str.add(")");

    if (returnType) {
      str.add(" -> " + returnType.ident.originalName);
    }

    return true;
  }
}
