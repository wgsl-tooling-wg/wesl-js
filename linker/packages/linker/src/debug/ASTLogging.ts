import { AbstractElem, ModuleElem } from "../AbstractElems.ts";
import { importToString } from "./ImportToString.ts";
import { LineWrapper } from "./LineWrapper.ts";

export function astTree(elem: AbstractElem, indent = 0): string {
  const { kind, contents } = elem as ModuleElem;
  const str = new LineWrapper(indent);
  str.add(kind);
  addElemFields(elem, str);
  let childStrings: string[] = [];
  if (contents) {
    childStrings = contents.map(e => astTree(e, indent + 2));
  }
  if (childStrings.length) {
    str.nl();
    str.addBlock(childStrings.join("\n"), false);
  }

  return str.result;
}

export function elemToString(elem: AbstractElem): string {
  const { kind } = elem as ModuleElem;
  const str = new LineWrapper();
  str.add(kind);
  addElemFields(elem, str);
  let childStrings: string[] = [];
  if (childStrings.length) {
    str.nl();
    str.addBlock(childStrings.join("\n"), false);
  }
  return str.result;
}

function addElemFields(elem: AbstractElem, str: LineWrapper): void {
  addTextFields(elem, str) ||
    addVarishFields(elem, str) ||
    addStructFields(elem, str) ||
    addNameFields(elem, str) ||
    addFnFields(elem, str) ||
    addAliasFields(elem, str) ||
    addImport(elem, str) ||
    addRefIdent(elem, str) ||
    addDeclIdent(elem, str);
}

function addVarishFields(
  elem: AbstractElem,
  str: LineWrapper,
): true | undefined {
  const { kind } = elem;
  if (
    kind === "var" ||
    kind === "gvar" ||
    kind === "const" ||
    kind === "override"
  ) {
    str.add(" " + elem.name.ident.originalName);
    if (elem.typeRef) {
      str.add(":" + elem.typeRef.ident.originalName);
    }
    return true;
  }
}

function addTextFields(
  elem: AbstractElem,
  str: LineWrapper,
): true | undefined {
  if (elem.kind === "text") {
    const { srcModule, start, end } = elem;
    str.add(` '${srcModule.src.slice(start, end)}'`);
    return true;
  }
}

function addRefIdent(elem: AbstractElem, str: LineWrapper): true | undefined {
  if (elem.kind === "ref") {
    str.add(" " + elem.ident.originalName);
    return true;
  }
}

function addDeclIdent(elem: AbstractElem, str: LineWrapper): true | undefined {
  if (elem.kind === "decl") {
    str.add(" %" + elem.ident.originalName);
    return true;
  }
}

function addAliasFields(
  elem: AbstractElem,
  str: LineWrapper,
): true | undefined {
  if (elem.kind === "alias") {
    const { name, typeRef } = elem;
    const prefix = name.ident.kind === "decl" ? "%" : "";
    str.add(" " + prefix + name.ident.originalName);
    str.add("=" + typeRef.ident.originalName);
    return true;
  }
}

function addStructFields(
  elem: AbstractElem,
  str: LineWrapper,
): true | undefined {
  if (elem.kind === "struct") {
    str.add(" " + elem.name.ident.originalName);
    return true;
  }
}

function addImport(elem: AbstractElem, str: LineWrapper): true | undefined {
  if (elem.kind === "import") {
    str.add(" " + importToString(elem.imports));
    return true;
  }
}

function addNameFields(
  elem: AbstractElem,
  str: LineWrapper,
): true | undefined {
  if (elem.kind === "name") {
    str.add(" " + elem.name);
    return true;
  }
}

function addFnFields(elem: AbstractElem, str: LineWrapper): true | undefined {
  if (elem.kind === "fn") {
    const { name, params, returnType } = elem;
    str.add(" " + name.ident.originalName);

    str.add("(");
    const paramStrs = params
      .map(p => p.name.ident.originalName + ": " + p.typeRef.ident.originalName)
      .join(", ");
    str.add(paramStrs);
    str.add(")");

    if (returnType) {
      str.add(" -> " + returnType.ident.originalName);
    }

    return true;
  }
}
