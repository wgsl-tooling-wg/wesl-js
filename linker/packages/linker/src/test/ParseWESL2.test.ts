import { expect, test } from "vitest";
import { astTree } from "../ASTLogging.ts";
import { parse2Test } from "./TestUtil.ts";

test("link global var", () => {
  const src = `var x: i32 = 1;`;
  const ast = parse2Test(src);
  const result = astTree(ast.rootModule);
  expect(result).toMatchInlineSnapshot(`
    "module
      var x:i32
        text 'var '
        ident %x
        text ': '
        ident i32
        text ' = 1'
      text ';'"
  `);
});
