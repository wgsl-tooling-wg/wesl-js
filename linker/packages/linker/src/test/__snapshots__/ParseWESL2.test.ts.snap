// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`parse unicode ident 1`] = `
"module
  text '
  '
  fn Δέλτα()
    text 'fn '
    decl %Δέλτα
    text '(){}'
  text ' 
  '
  fn réflexion()
    text 'fn '
    decl %réflexion
    text '(){}'
  text ' 
  '
  fn Кызыл()
    text 'fn '
    decl %Кызыл
    text '(){}'
  text ' 
  '
  fn 𐰓𐰏𐰇()
    text 'fn '
    decl %𐰓𐰏𐰇
    text '(){}'
  text ' 
  '
  fn 朝焼け()
    text 'fn '
    decl %朝焼け
    text '(){}'
  text '
  '
  fn سلام()
    text 'fn '
    decl %سلام
    text '(){}'
  text ' 
  '
  fn 검정()
    text 'fn '
    decl %검정
    text '(){}'
  text ' 
  '
  fn שָׁלוֹם()
    text 'fn '
    decl %שָׁלוֹם
    text '(){}'
  text '
  '
  fn गुलाबी()
    text 'fn '
    decl %गुलाबी
    text '(){}'
  text '
  '
  fn փիրուզ()
    text 'fn '
    decl %փիրուզ
    text '(){}'
  text '
  '"
`;
